Add and refactor mock-connector from knot-service-source to work with new AMQP API

CaioCordeiro opened the pull request:
Changes:

- Add mock-connector from knot-service-source
- Rename it to 'mock-cloud'
- Refactor to work with the new AMQP API
- Added a script to create a rabbitmq docker enviroment

CaioCordeiro committed:
Add mock-connector from knot-service-source

Add mock-connector script from knot-service-source.
This file is used to mock the cloud infrastructure for testing.
It's originally from https://github.com/CESARBR/knot-service-source/tree/5410478/test
Commit hash: 36655a6133549e442e910c6565b3ea828ab4d957

CaioCordeiro committed:
Rename mock-connector

Rename mock-connector.py to mock-cloud.py to give more clarity to the
functionality of the script.

luigiluz commented on test/mock-cloud.py:
```diff
@@ -24,209 +24,183 @@
+    # Device exchange
+    channel.exchange_declare(exchange=device_exchange, durable=True,
+                             exchange_type='direct')
+    # Data reciving exchange
```
```suggestion
    # Data receiving exchange
```

luigiluz commented on test/mock-cloud.py:
```diff
@@ -24,209 +24,183 @@
     return channel
 
 # Parser sub-commands
+
+
```
Are these two blank lines necessary?

CaioCordeiro replied:
Not at all, i can remove them.

luigiluz reviewed (changes requested):
Nice PR, Congrats!
I've made two remarks in respect to comments and blank lines, but the code looks good!
If it is possible, I would just consider splitting commit [dc87b34](https://github.com/CESARBR/knot-cloud-sdk-c/pull/2/commits/dc87b34c4053bc12559904b324c82a7d6ea77d33) in more than one commit. 
E.g: One to remove the request and update functionalities, another one to change the EVENT names

k-cabral commented on test/test_rabbitmq_env.sh:
```diff
@@ -0,0 +1,3 @@
+docker network create myapp_net
+
+docker run -d --hostname rabbitmqhost --name rabbitmq -p 15672:15672 -p 5672:5672 rabbitmq:3-management
```
I believe the `rabbitmq:management` image might be overkill to most of the use cases, I'd suggest using a smaller and simpler one by default, such as `rabbitmq:alpine`.

CaioCordeiro replied:
As i discussed with you, @alla3, @vitbaq  and @luigiluz , i opted to remove this file, add this simple script to the README and change the docker image from  `rabbitmq:management`  to `rabbitmq:management-alpine` which is way lighter.
This changes can be seen in the commit:  8c50aac

k-cabral commented on test/mock-cloud.py:
```diff
@@ -24,39 +24,39 @@
 
 MESSAGE_EXPIRATION_TIME_MS = '10000'
-
+# Babeltower Events API - v2.0.0
```
Nit: I like to avoid adding version numbers in code because they tend to get outdated somewhat quickly.
Maybe, because this is an API, it might be a fine, but I'd still at least remove the minor and patch version numbers, since these changes won't break compatibility.

k-cabral reviewed (changes requested):
Good job! I have some change requests and some minor nitpicks:

- I definitely think 6eeb007 needs to be split into up to three separate commits. Although it's relativelly small, it does many different things.
- I was a little confused by ad12e93's commit description, so here's a suggestion:

> Add mock-connector script from knot-service-source.
> This file is used to mock the cloud infrastructure for testing.
> It's originally from https://github.com/CESARBR/knot-service-source/tree/5410478/test

- Nit: On 6e3b4ac, it'd be cool to explain why those features are being removed in the first place.

CaioCordeiro committed:
Remove the request and update functionalities

This patch removes the 'request' and 'update' functionalities.
These features are very specific and outside the mock-cloud's
desired scope.

CaioCordeiro committed:
Update README to include the mock-cloud docs

This patch adds a documentation of how to execute the mock-cloud as a
test environment.

k-cabral commented on test/mock-cloud.py:
```diff
@@ -158,39 +158,47 @@ def __on_msg_received(args, channel, method, properties, body):
+        exchange=device_exchange, queue=queue_name, routing_key=EVENT_REGISTER)
     channel.queue_bind(
-        exchange=cloud_exchange, queue=queue_name, routing_key='device.cmd.list')
+        exchange=device_exchange, queue=queue_name, routing_key=EVENT_UNREGISTER)
```
I'm not sure what's the PEP8 recommendation for long lines or breaking long lines, nor if this is appropriate for Python. Might be worth double checking.

k-cabral commented on test/mock-cloud.py:
```diff
@@ -24,39 +24,39 @@
-
-EVENT_DATA = 'data.publish'
+EVENT_SCHEMA = 'device.schema.sent'
+KEY_SCHEMA = 'device.schema.updated'
```
Does it make sense to define the key and event names according to the exchange name? How likely is it for the exchange name to change?
e.g.: `EVENT_AUTH = "%s.auth" % device_exchange`

k-cabral reviewed (approved):
LGTM, I believe this is mergeable.
That being said, I've left some comments. None of them are required, but they might be worth a thought. On the same tone, (nit:) I'd suggest applying f648165b06f299827153bd29ab625ca3c9987a94 before updating/refactoring the code, which means before 8004f91bb817257e3fa2f5e0bd5ac1764515391c.

CaioCordeiro committed:
Update EVENTs responses

This patch updates the EVENTs responses to the AMQP connections to work
with the new AMQP API.

vitbaq reviewed (changes requested):
Put this commit (31b5ab84df5da1633123760ef09b44e92c4033b2) before this one (58e1c66b6a32c3b2d64a60fdd105d83199973088)

We must have the code working on every commit (not breaking). For example, you exclude the fog_exchange and cloud_exchange but still using it on the commit 58e1c66b6a32c3b2d64a60fdd105d83199973088

Changing the order of the commits you can avoid this problem. You can evolve the code, it doesn't have to be functional, but it can't break.

CaioCordeiro committed:
Bind connection to the new exchanges

This patch binds the mock-cloud AMQP connection to the new exchanges
required by each of the EVENTs from the API.

CaioCordeiro committed:
Update EVENTs names

This patch updates the EVENTs names to work with the new AMQP API.

k-cabral reviewed (commented):
I tried to revise the PR but I've encountered an error when trying to run the program:
```
File "/usr/local/lib/python3.8/dist-packages/pika/adapters/blocking_connection.py", line 450, in _create_connection
    raise self._reap_last_connection_workflow_error(error)
pika.exceptions.AMQPConnectionError
Traceback (most recent call last):
  File "test/mock-cloud.py", line 201, in <module>
    options.func(options)
  File "test/mock-cloud.py", line 155, in msg_consume
    channel = __amqp_start()
  File "test/mock-cloud.py", line 141, in __amqp_start
    connection = pika.BlockingConnection(
  File "/usr/local/lib/python3.8/dist-packages/pika/adapters/blocking_connection.py", line 359, in __init__
    self._impl = self._create_connection(parameters, _impl_class)
  File "/usr/local/lib/python3.8/dist-packages/pika/adapters/blocking_connection.py", line 450, in _create_connection
    raise self._reap_last_connection_workflow_error(error)
pika.exceptions.AMQPConnectionError
```
I don't think it's any issue with your code, though, so it would be cool if someone else could try it too. Is it running fine for you, Caio?
